Block Comments: Highlight the related block

karthick-murugan opened the pull request:
**What?**

Closes #71244

This PR introduces a visual highlight for the related block when a comment is selected in the sidebar. It ensures that users can quickly identify which block a comment is associated with inside the editor.

**Why?**

Currently, when comments are shown in the sidebar, there is no clear indication in the canvas of which block they belong to. This creates confusion and slows down workflows for authors and reviewers.

By highlighting the related block whenever its associated comment is selected, users will have a clear visual indicator of context, improving the overall commenting experience.

**How?**

When a comment is selected:

The related block is programmatically selected and a temporary data-comment-highlight="true" attribute is applied to the block’s DOM element inside the editor iframe.

A CSS rule applies a subtle highlight style to the block (e.g., outline or background fade).

When the comment is deselected or another block/comment is selected:

The data-comment-highlight attribute is removed, restoring the block to its normal appearance.

**Implementation details**

Uses iframe.contentDocument.querySelector('[data-block="..."]') to locate the block.

Adds data-comment-highlight="true" instead of class manipulation, since class selectors weren’t applying reliably inside the editor iframe.

Adds scoped CSS targeting [data-comment-highlight="true"] to visually differentiate highlighted blocks.

**Testing Instructions**

Open a post or page in the block editor.

Add a few blocks and insert comments linked to those blocks.

From the sidebar, click on a comment.

✅ The related block in the editor should receive a visible highlight.

**Screenshot/ Video** 

https://github.com/user-attachments/assets/31eb64cb-f74f-4151-82e2-827a367ddd76

karthick-murugan committed:
Image size fix in lightbox

karthick-murugan committed:
Revert "Image size fix in lightbox"

This reverts commit 63f81c1e9cbb3b0513abca2b8cb579fef374ed7b.

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Merge branch 'WordPress:trunk' into trunk

karthick-murugan committed:
Highlight and scroll to related blocks

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,25 +48,73 @@ export function Comments( {
+			selectBlock( relatedBlocks[ 0 ] );
+
+			const iframe = document.querySelector(
+				'iframe[name="editor-canvas"]'
```
selecting and manipulating elements in the DOM this way feels fragile. Did you explore using a block attribute / datastore to tie the comment to the block/highlighting? 

Perhaps we could leverage the existing mechanism that ties between blocks and comments - clicking on a block highlights the related comment thread?

karthick-murugan replied:
@adamsilverstein - Updated the feedback. 


https://github.com/user-attachments/assets/d2bff01b-6eac-49eb-b6d7-aba20d86840e

karthick-murugan committed:
Feedback changes updated

adamsilverstein commented:
@karthick-murugan - thanks for the updates! this looks much better, I will review the code shortly.

I noticed a few behaviors that raised questions or feel glitchy that we may want to address:

* Selection highlighting worked sometimes and sometime did not, not sure yet under what conditions
* If I select a comment and the related block is off screen, I can't see the highlighting. Should we have the browser scroll to the right position in the document so the block is visible?  Similarly, should we scroll to he highlighted comment in the comment sidebar when a block is selected? if there are many/long comments this could easily be off-screen.
* When I select multiple blocks, I noticed no comments are selected. Is that what we want?
* Clicking on another should un-highlight/deselect the previous selection (steps to reproduce 1. select a comment, highlighting the block. 2. click another comment in the sidebar. The comment opens but the old comment and block are still highlighted.
* Linking doesn't work as expected if the post isn't saved (to reproduce - add some comments to some blocks and test the linking without saving.
*

karthick-murugan committed:
Latest Feedback updated

adamsilverstein commented:
Thanks for the new updates @karthick-murugan, I will give this a test and review!

> Regarding the case of multiple block selection: currently, no comments are highlighted when more than one block is selected. Could you clarify how you’d like to handle this scenario? Should we highlight multiple comments corresponding to the selected blocks, or is there another preferred approach?

I'm not really sure how this should be handled, my hunch is we should highlight all related. We don't need to address this in the current PR however - after testing how Google docs handles this, I will open a follow up Issue for the task.

karthick-murugan commented:
> @karthick-murugan this worked great in my testing on a **page**, however it doesn't seem to work in **posts** at all. Not sure why yet, I will investigate further.

@adamsilverstein - Tested the functionality in **Posts** and it is working fine for me. Please find the attached video taken from Posts.


https://github.com/user-attachments/assets/331c1bf0-e371-46e2-baee-c0ab3df965c9

adamsilverstein commented:
>  Hmmm, I will re-test- perhaps I have something locally or on core trunk interfering.

This seems to be on my dev/trunk install only. Trying a fresh install of WordPress, I don't see the issue.

On my dev install, it worked for me on posts after changing my theme to Twenty-Twenty-Four. Switching back to Twenty-Twenty-Five, I see the issue again on posts. the selection isnt triggered. I even tried resetting my db and reinstalling the theme.

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,25 +48,184 @@ export function Comments( {
-				: null,
-		};
-	}, [] );
+	// inside Comments()
```
Can you expand this documentation line or remove it?

karthick-murugan replied:
Updated.

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,25 +48,184 @@ export function Comments( {
-				? getBlockAttributes( _clientId )?.blockCommentId
-				: null,
-		};
-	}, [] );
```
this and the next three lines look like whitespace only changes that can be removed unless the linter is dictating this formatting.

karthick-murugan replied:
@adamsilverstein - The linter needs that format.

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,25 +48,184 @@ export function Comments( {
 		showCommentBoard && blockCommentId ? blockCommentId : null
 	);
 
-	const clearThreadFocus = () => {
```
this change seems unrelated to the PR

karthick-murugan replied:
Updated

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -113,7 +272,10 @@ export function Comments( {
 							onCommentResolve={ onCommentResolve }
 							onEditComment={ onEditComment }
 							isFocused={ focusThread === thread.id }
-							clearThreadFocus={ clearThreadFocus }
```
the change to clearThreadFocus seems unrelated to the PR

karthick-murugan replied:
Updated

karthick-murugan committed:
Feedback changes updated

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +48,190 @@ export function Comments( {
+			)?.attributes;
+			if ( blockAttributes?.blockCommentId ) {
+				setFocusThread( blockAttributes.blockCommentId );
+				const commentScrollTimeout = setTimeout( () => {
```
is this setTimeout required?

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +48,190 @@ export function Comments( {
+				scrollTimeouts.current.push( commentScrollTimeout );
+			}
+		} catch {
+			const retryTimeout = setTimeout( () => {
```
can we remove this timeout?

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +48,190 @@ export function Comments( {
+	const { toggleBlockHighlight } = useDispatch( blockEditorStore );
+
+	useEffect( () => {
+		const reset = () => {
```
I noticed we have `useScrollIntoView` in packages/block-editor/src/components/block-list/use-block-props/use-scroll-into-view.js - can you take a look at that? maybe we can use it directly here? it offers some nice features including honoring the users prefers reduced motion setting for the scrolling behavior.

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +48,190 @@ export function Comments( {
+
+	useEffect( () => {
+		const reset = () => {
+			if ( retryRef.current ) {
```
I tested removing retryRef and prevRef and didn't notice any difference - hence my question about what these do.

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +48,190 @@ export function Comments( {
+		return reset;
+	}, [ selectedBlockClientId, toggleBlockHighlight, blocks ] );
+
+	// Function to find and select blocks by comment ID
```
Inline comments like this should end in a period.

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +48,190 @@ export function Comments( {
+			prevRef.current = selectedBlockClientId;
+
+			// Scroll block to center
+			const blockElement = document.querySelector(
+				`[data-block="${ selectedBlockClientId }"]`
+			);
```
Let's avoid this approach. `document.querySelector` won't work because the block is inside the iframe editor instance. This code means we select the list view item, not the block.

karthick-murugan replied:
@t-hamano - Updated

t-hamano reviewed (commented):
Sorry for the sudden message. I think it would be a good idea to carefully consider what is the best approach first.

I think what's expected here is that selecting a comment doesn't actually select the block, but just "visually highlight" it. Imagine a keyboard user clicking the Edit button to edit a comment. The user just wants to edit the comment, but the focus moves to the block.

https://github.com/user-attachments/assets/486f1dbc-c941-4365-b9b5-79fb5697b57e

Perhaps from an accessibility perspective, we might want to consider how focus should move when we click on a comment and how a screen reader should announce it.

cc @joedolson 

Another important point is that blocks are rendered in the iframed editor, so `document.querySelector` cannot find them.

t-hamano commented:
> I think what's expected here is that selecting a comment doesn't actually select the block, but just "visually highlight" it. Imagine a keyboard user clicking the Edit button to edit a comment. The user just wants to edit the comment, but the focus moves to the block.

I am wondering if the following approach is possible:

- Expose the `useBlockElement()` hook as a private api fomr the `@wordpress/block-editor` package
- Find a related block when a comment is clicked
- Get the target block's ref by using the `useBlockElement` hook and the found clientId
- Use `scrollIntoView()` to scroll so that the found block is visible.
- Dispatch `toggleHighlight()` so that the blue outline is rendered 

cc @ellatrix @Mamaduka for technical feedback

joedolson commented:
Thanks for the ping! 

When a comment is clicked on, the comment should receive focus. Moving focus somewhere else would definitely be confusing. When you click on an input, the input you clicked on should receive focus - anything else would be very disorienting. A link is expected to take you elsewhere. Screen reader and keyboard users will generally not perform any kind of action on other types of elements.

Different cues are for different audiences; we want a visual highlight to offer sighted users an affordance about how the comment relates to other content on the page.

Screen reader & keyboard users need to have a mechanism to easily navigate from a comment to the block it's related to.

karthick-murugan committed:
Feedback changes added

t-hamano commented on packages/block-editor/src/components/index.js:
```diff
@@ -115,6 +115,7 @@ export { default as __experimentalUseResizeCanvas } from './use-resize-canvas';
 export { default as BlockInspector } from './block-inspector';
 export { default as BlockList } from './block-list';
 export { useBlockProps } from './block-list/use-block-props';
+export { useBlockElement } from './block-list/use-block-props/use-block-refs';
```
This means that the `useBlockElement` hook is exposed and available to consumers, but we should avoid it for now. It should be exported as a private API.

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +51,133 @@ export function Comments( {
+				block: 'center',
+			} );
+		}
+	}, [ highlightedBlockElement ] );
```
All we want to achieve here is to simply find the block and scroll to it when a comment thread click is triggered. Why do we need these effects? Can't we do all of this in `handleCommentSelect`?

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,27 +51,133 @@ export function Comments( {
 	);
 
+	// Track the currently highlighted block
+	const [ highlightedBlockId, setHighlightedBlockId ] = useState( null );
```
If I understand correctly, when you perform the `toggleBlockHighlight` action, the highlighted other block will be automatically removed, so we shouldn't need to keep track of which block is highlighted.

karthick-murugan committed:
Feedback changes

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -95,16 +148,13 @@ export function Comments( {
 						className={ clsx(
 							'editor-collab-sidebar-panel__thread',
 							{
-								'editor-collab-sidebar-panel__active-thread':
```
why are these lines removed, can we keep them?

karthick-murugan replied:
Updated

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -95,16 +148,13 @@ export function Comments( {
-									blockCommentId &&
-									blockCommentId === thread.id,
 								'editor-collab-sidebar-panel__focus-thread':
-									focusThread && focusThread === thread.id,
```
this line shouldn't need to be changed

karthick-murugan replied:
Updated

t-hamano commented on packages/block-editor/src/components/index.js:
```diff
@@ -115,6 +115,7 @@ export { default as __experimentalUseResizeCanvas } from './use-resize-canvas';
 export { default as BlockInspector } from './block-inspector';
 export { default as BlockList } from './block-list';
 export { useBlockProps } from './block-list/use-block-props';
+export { useBlockElement as __unstableUseBlockElement } from './block-list/use-block-props/use-block-refs';
```
Even if you add the `__unstable` prefix, it is not a private API. We need to import the `useBlockElement` hook [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/private-apis.js) and lock it.

Then, use it in code like this:

```javascript
/**
 * WordPress dependencies
 */
import { privateApis as blockEditorPrivateApis } from '@wordpress/block-editor';

/**
 * Internal dependencies
 */
import { unlock } from '../lock-unlock';

const { useBlockElement } = unlock( blockEditorPrivateApis );
```

karthick-murugan replied:
Updated

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,25 +51,75 @@ export function Comments( {
+	const { toggleBlockHighlight } = useDispatch( blockEditorStore );
+
+	// State to track the highlighted block for scrolling
+	const [ highlightedBlockId, setHighlightedBlockId ] = useState( null );
```
This can be deleted, right?

karthick-murugan replied:
Updated

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -48,25 +51,75 @@ export function Comments( {
+
+	// Effect to handle scrolling when highlighted block element changes
+	useEffect( () => {
+		if ( highlightedBlockElement ) {
+			highlightedBlockElement.scrollIntoView( {
+				behavior: 'smooth',
+				block: 'center',
+			} );
+		}
```
I think we can move this process into the `handleCommentSelect`.

karthick-murugan replied:
@t-hamano Thanks for the suggestion! I tried moving the scrolling logic into handleCommentSelect, but it doesn't work due to `useBlockElement` is a hook that can be called at the component's top level

t-hamano replied:
This effect causes an unintentional behaviour. When moving your mouse over the list view, the canvas automatically scrolls. I'll try to find a way to remove this effect. 


https://github.com/user-attachments/assets/d8b4c045-f318-4bd2-aaa3-3aef2987800b

karthick-murugan committed:
Feedback changes

t-hamano committed:
Merge branch 'trunk' into highlight-related-block

t-hamano committed:
Remove unecessarunnecessary effect and new api

adamsilverstein commented:
This is looking really good @t-hamano - the code is much more succinct and easier to understand. Nice work!

I noticed one small behavior that leaves two comments highlighted and feels a little off. 

To reproduce:

1. show comments and leve a few test comments
2. click a block with or without a comment which highlights the related comment
3. click a different comment, now both are highlighted

Can we make sure only one comment is highlighted (the one you clicked on last)?

### Screenshots
existing comment
<img width="307" height="600" alt="nultiple comments highlighted" src="https://github.com/user-attachments/assets/ada9be71-ef56-4bcc-8afe-1b2dfb7e5eb4" />

new comment
<img width="294" height="668" alt="comment plus select" src="https://github.com/user-attachments/assets/bfe5678f-ab5d-4ae9-823b-7ac77c94cd2f" />

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -17,15 +17,21 @@ import {
 import CommentAuthorInfo from './comment-author-info';
 import CommentForm from './comment-form';
 
+const { useBlockElement } = unlock( blockEditorPrivateApis );
```
Nice!

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -129,9 +123,63 @@ function Thread( {
+
+	const { toggleBlockHighlight } = useDispatch( blockEditorStore );
+
+	// Find first block that have this comment ID - run at component root level.
```
```suggestion
	// Find first block that has this comment ID - run at component root level.
```

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -129,9 +123,63 @@ function Thread( {
+		onFocusThread( threadId );
+		if ( relatedBlock && relatedBlockElement ) {
+			relatedBlockElement.scrollIntoView( {
+				behavior: 'smooth',
```
I believe this scroll behavior should respect the user agent's "preferred reduced motion" preferences, similar to https://github.com/WordPress/gutenberg/blob/aa3b96150333d9102fe3bbb07cff3041fda69a3b/packages/block-editor/src/components/block-list/use-block-props/use-scroll-into-view.js#L22

cc: @joedolson for accessibility guidance here

t-hamano replied:
Good catch, we should use the `useReducedMotion` hook.

adamsilverstein reviewed (commented):
Excellent! Left some small feedback.

t-hamano commented:
> Can we make sure only one comment is highlighted (the one you clicked on last)?

Sorry, my commit might have introduced this behaviour. This should be fixed before shipping.

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -129,9 +123,63 @@ function Thread( {
+	const { toggleBlockHighlight } = useDispatch( blockEditorStore );
+
+	// Find first block that have this comment ID - run at component root level.
+	const relatedBlock = useMemo( () => {
+		if ( ! thread.id || ! blocks ) {
+			return null;
+		}
+		const findFirstBlock = ( blockList ) => {
+			for ( const block of blockList ) {
+				if ( block.attributes?.blockCommentId === thread.id ) {
+					return block.clientId;
+				}
+				if ( block.innerBlocks ) {
+					const found = findFirstBlock( block.innerBlocks );
+					if ( found ) {
+						return found;
+					}
+				}
+			}
+			return null;
+		};
+
+		return findFirstBlock( blocks );
+	}, [ thread.id, blocks ] );
```
Note for the future (not a blocker): We need to find a better way to derive this value; the `blocks` will update often when the user is typing, which can affect performance.

Unfortunately, we can't do this inside `handleCommentSelect` due to the `useBlockElement` nature.

karthick-murugan replied:
@Mamaduka - Please let us know how can we move forward with this issue. Do we need an updated approach of finding the Block?

Mamaduka replied:
As I mentioned, it’s not a blocker for this PR. Just a note (mostly for my self) for future refactoring.

Yes, current block client ID look up isn’t efficient.

t-hamano commented:
@karthick-murugan, can you first resolve the conflicts so that this PR can be tested in a compatible state with the trunk branch?

karthick-murugan committed:
Merge branch 'trunk' into highlight-related-block

karthick-murugan committed:
Fixed feedback issues

karthick-murugan commented:
Fixed the previously selected comment highlight issue and bi-directional highlighting. Please refer the latest video here.

https://github.com/user-attachments/assets/42cf7081-6465-42cc-9780-30669cc78227

adamsilverstein commented:
> Can we use flashBlock instead of highlightBlock? 

This sounds good to me - @karthick-murugan can you give this a try?

adamsilverstein commented:
> Fixed the previously selected comment highlight issue and bi-directional highlighting. Please refer the latest video here.

This tested well for me @karthick-murugan - one small issue I noticed was clicking on a comment to highlight a block, then clicking on a block without a comment, the first highlighted block remained highlighted.  I assume switching to `flashBlock` will fix this anyway.

karthick-murugan committed:
Added flashblock and updated highlight issues

t-hamano committed:
Update packages/editor/src/components/collab-sidebar/comments.js

Co-authored-by: Adam Silverstein <adamsilverstein@earthboundhosting.com>

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,25 +78,129 @@ export function Comments( {
+	] );
+
+	// Clear highlights when switching to blocks without comments
+	useEffect( () => {
```
What does this effect do? I think it will work fine without it.

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,25 +78,129 @@ export function Comments( {
+			}
+		}
+
+		// Clear the add comment form when selecting a comment.
```
```suggestion
		// Clear the add comment form when selecting a comment.
```

Now we're using `flashBlock`, so I think we don't need  `toggleBlockHighlight`.

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,25 +78,129 @@ export function Comments( {
+			relatedBlockClientId &&
+			relatedBlockClientId !== selectedBlockClientId
+		) {
+			toggleBlockHighlight( relatedBlockClientId, false );
```
```suggestion
				: null;
```

Now we're using `flashBlock`, so I think we don't need  `toggleBlockHighlight`.

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,25 +78,129 @@ export function Comments( {
+		// Flash block when comment is focused (non-persistent highlight).
+		if ( relatedBlockClientId ) {
+			flashBlock( relatedBlockClientId );
+		}
```
Can't this be done inside `handleCommentSelect`?

karthick-murugan replied:
@t-hamano - Updating this change the comment scrolling is not working. Other than this all changes are updated.

karthick-murugan committed:
Feedback changes

t-hamano commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,25 +78,95 @@ export function Comments( {
+		// Flash block when comment is focused (non-persistent highlight).
+		if ( relatedBlockClientId ) {
+			flashBlock( relatedBlockClientId );
+		}
```
I think we should put this code into the `handleCommentSelect`. My understanding is that there is no need to flash blocks when they are selected, because the user already knows which block they selected. The only time you flash a block is when a comment is clicked. No two-way sync is required.

karthick-murugan replied:
@t-hamano - Updated all the changes. Thanks for reviewing.

karthick-murugan committed:
Feedback changes updated

adamsilverstein commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,27 +84,60 @@ export function Comments( {
+			// Set the focused thread.
+			setFocusThread( threadId );
+
+			// Flash the related block when comment is clicked
```
```suggestion
			// Flash the related block when a comment is clicked.
```

t-hamano committed:
Update packages/editor/src/components/collab-sidebar/comments.js

Co-authored-by: Adam Silverstein <adamsilverstein@earthboundhosting.com>

t-hamano reviewed (commented):
> Changing the scrolling from "smooth" to "instant" fixed this in my local testing. what do you think?

I prefer this approach. For example, when you select a block from the list view, the behavior is "instant" instead of "smooth."

karthick-murugan committed:
Flashblock and scroll updates added

karthick-murugan commented:
> @karthick-murugan this tested well, all of the previous issues seem resolved. I noticed a few small things still:
> 
> * if the related block is far off the current screen (for example with a long document), the highlighting flash takes place before the scrolling reaches the highlighted block so I never see it. Changing the scrolling from "smooth" to "instant" fixed this in my local testing. what do you think?
> * the flashBlock effect seems rather fast for me, I can see how it works for copying where your visual focus is already on the block and the highlighting is more of a confirmation that your action did something. Here it feels too fast: I barely noticed it at first. I'm curious for some design feedback on this question. For something longer, we could extend flashBlock to accept a timeout value or create a highlightBlock analog.
> * If we do keep the 'smooth' scrolling, we need to address user's reduced motion settings.

@adamsilverstein @t-hamano - I’ve addressed the scroll behavior which now uses instant scrolling to ensure the highlight is visible, and I’ve adjusted the flash timing so the effect is more noticeable. 


https://github.com/user-attachments/assets/f7203353-1062-418a-bf85-30e7363d97dd

karthick-murugan committed:
Removing FlashBlock API changes

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,27 +84,60 @@ export function Comments( {
 		setShowCommentBoard( false );
 	};
 
+	const [ focusThread, setFocusThread ] = useState(
+		showCommentBoard && blockCommentId ? blockCommentId : null
+	);
+
+	// Sync comment ↔ block highlighting bidirectionally.
+	useEffect( () => {
+		// Highlight comment when block is selected.
+		if ( blockCommentId && ! focusThread ) {
+			setFocusThread( blockCommentId );
+		}
+	}, [ blockCommentId, focusThread, blocks, setFocusThread ] );
```
Similar sync effects can be tricky. I know there are "checks" in place, but this effect will run a lot of times. Each block update will trigger the effect.

IMO, we have a clear use, even "click on a comment element", there should be no need for any side-effects. Technically, we should be able to house all required logic in `handleCommentSelect`.

karthick-murugan replied:
@Mamaduka – Removing the useEffect will stop highlighting comments when any block is selected, but the block flashing when a comment is selected will continue to work as expected. Do you think it’s fine if I proceed with this approach?

Mamaduka replied:
It's not a big blocker; we can start here and refactor later.

I just noticed that the `blocks` dependency isn't required for this effect; what was the reason for adding it?

karthick-murugan replied:
Its just for syncing the highlight.

Mamaduka replied:
Seems like an odd hack.

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -49,27 +84,60 @@ export function Comments( {
+			}
+		},
+		[ blocks, flashBlock, setFocusThread, setShowCommentBoard ]
+	);
```
Why do we need two `handleCommentSelect` functions? The logic should be co-located in a single callback below.

P.S. Memoization is sort of useless here. The `blocks` dependency will often change.

karthick-murugan replied:
@Mamaduka - Updated both the changes.

Mamaduka replied:
Sorry, I should've been clearer. What I was suggesting was moving all the logic into `handleCommentSelect` defined in the `Thread` component.

karthick-murugan replied:
@Mamaduka - Thanks for the clarification! I've moved all the logic into the handleCommentSelect function defined in the Thread component as you suggested.

Changes made:
✅ Removed the handleCommentSelect function from the main Comments component
✅ Consolidated all logic (clearing comment board, setting focus, flashing blocks, scrolling) into the single handleCommentSelect in the Thread component
✅ Removed the useEffect for comment ↔ block highlighting sync
✅ Removed unnecessary useCallback memoization

The code is now much cleaner with a single source of truth for comment selection logic, and no performance issues from the frequently changing blocks dependency in memoization.

Mamaduka reviewed (commented):
Left a couple more notes. I think we're close to finishing this.

karthick-murugan committed:
Remove duplicate handleCommentSelect

karthick-murugan committed:
handleCommentSelect to Thread Component

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -134,9 +153,57 @@ function Thread( {
+		// Set the focused thread.
+		setFocusThread( threadId );
+
+		// Flash the related block when a comment is clicked.
+		if ( blocks ) {
+			const relatedBlockClientId = findBlockByCommentId(
+				threadId,
+				blocks
+			);
+			if ( relatedBlockClientId ) {
+				// Use longer timeout for comment-to-block highlighting (800ms)
+				flashBlock( relatedBlockClientId, 800 );
+			}
+		}
```
We can just use the `relatedBlock` value here. There's no need to find the same value twice.

karthick-murugan replied:
@Mamaduka - Yes done.

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -107,36 +101,34 @@ export function Comments( {
-		if ( blockCommentId && ! focusThread ) {
-			setFocusThread( blockCommentId );
-		}
-	}, [ blockCommentId, focusThread, blocks, setFocusThread ] );
```
I thought we were keeping this effect for now? Was the removal intentional?

Sorry, just saw this mentioned in https://github.com/WordPress/gutenberg/pull/71308#discussion_r2362313080. But it's still not clear if the removal was intentional.

Related: https://github.com/WordPress/gutenberg/pull/71308#discussion_r2362163921

karthick-murugan replied:
@Mamaduka - Not intentional. Added back.

karthick-murugan committed:
Updating feedback

Mamaduka commented on packages/editor/src/components/collab-sidebar/comments.js:
```diff
@@ -134,9 +160,50 @@ function Thread( {
+				behavior: 'instant',
+				block: 'center',
+			} );
+		}
```
Maybe we should combine these so that they can "flash" the block highlight after the scroll.

P.S. Most of the inline comments inside the `handleCommentSelect` function aren't needed, IMO. What each part of the code is doing is clear.

karthick-murugan replied:
@Mamaduka - Good point! I've combined the flash and scroll so the block highlights after scrolling, and remove the unnecessary inline comments since the code is self-explanatory. Thank You.

karthick-murugan committed:
FlashBlock after scroll

t-hamano reviewed (approved):
LGTM! Thank you so much for sticking with it until the end. I believe this is the best we can do for now. If necessary, let's follow up by removing unnecessary effects, improving performance, and refactoring.
